Auto set first Navigation Menu as current if only 1 exists

getdave opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->



## What?

This PR causes the Navigation block to auto select the first Navigation Menu as its current active menu _if_ there is only a single `wp_navigation`.

## What is NOT

This PR does _not_:

- attempt to automatically provide Page List as a fallback if the block is empty (we need to keep the scope reasonable).
- [handle uncontrolled inner blocks saving](https://github.com/WordPress/gutenberg/pull/39883).



## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Currently an empty Nav block requires the user to select the Navigation Menu they want to use. However, most users will probably only have a single Navigation Menu - `Primary`.

On Theme switch currently the user is required to reselect their Menu each time which is unwanted overhead. The change in this PR means that if there is only a single Menu then the block will automatically use that Menu. This avoids the need to reselect the Menu.

Note that if there is more than 1 Menu then we don't pursue this behaviour as we cannot determine which Menu would be the most appropriate.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Check for how many Navigation Menus are available and if there is only one then auto-select that as the current Nav block's active Menu.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

- Delete all Navigation Menus.
- Insert Nav block.
- Check that default placeholder is shown with option to `Start empty`.
- Click Start Empty to create a new Nav Menu. Populate with some items you can easily remember. Publish to save the Menu.
- Start a new Post.
- Insert Nav block. 
- See that it will automatically select the Navigation Menu - because you have only 1 which you created above.
- Use the `Select menu` dropdown and click `Create empty menu`.
- See block resets to placeholder state.
- Click `Start empty`.
- Verify that a brand new empty menu is created.
- Publish to Save.
- Double check you have x2 Navigation Menus saved (see `/wp-admin/edit.php?post_type=wp_navigation`).
- Start a _new_ Post. 
- Insert the Nav block.
- Check that it does NOT automatically select a Navigation Menu. See default placeholder.

## Screenshots or screencast <!-- if applicable -->

https://user-images.githubusercontent.com/444434/161039198-1f0e0c5f-be81-4910-8020-4dd4ba69a78b.mov

getdave committed:
Auto set first Menu if exists

getdave committed:
Allow for user to explicitly create a new fully empty Menu

getdave committed:
Make code more self documenting

getdave committed:
Avoid flash of Placeholder

getdave committed:
Add test for auto selecting first Nav

Currently broken

getdave committed:
Add tests to cover auto selection of first Menus

getdave commented:
Tests are failing because of this line

https://github.com/WordPress/gutenberg/blob/4026bca3753cff9ac064079f30d61b40a00c8488/packages/e2e-test-utils/src/inserter.js#L171

Perhaps the Nav block is stealing focus when it automatically selects the first Navigation Menu?

getdave commented:
What's in the branch now should at least allow us to test how this feels as a user.

getdave committed:
Fix test issues

getdave committed:
Fix permissions test to account for auto selection

mtias commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -283,7 +286,31 @@ function Navigation( {
 		hasResolvedCanUserCreateNavigationMenu,
 	} = useNavigationMenu( ref );
 
+	// Setup initial block insertion state.
```
```suggestion
	// The block attempts to retrieve and prioritize any existing navigation menu unless
	// a specific ref is allocated or the user is explicitly creating a new menu. The aim is
	// for the block to "just work" from a user perspective using existing data.
```

mtias replied:
And I think we can remove the rest of the comment

getdave replied:
Updated.

mtias commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -283,7 +286,31 @@ function Navigation( {
+			return;
+		}
+
+		const firstAndOnlyNavigationMenu = navigationMenus[ 0 ];
```
this name seems a bit obscure, I'd suggest using `mainNavigation`

getdave replied:
Fair enough. I've decided to revert to having to var declared. Personal preference, but if it isn't going to be explicit then it's not self documenting.

scruffian commented:
This works as described. I can also confirm that if a theme uses an empty navigation block the first navigation is displayed

scruffian commented:
Although this does create an issue. If I have two navigation menus, one is displayed on the front end, but not in the site editor:

Frontend:
<img width="1440" alt="Screenshot 2022-04-01 at 12 51 24" src="https://user-images.githubusercontent.com/275961/161258104-28cd5e0d-7bc5-4071-9a00-e5690e386923.png">

Site Editor
<img width="1439" alt="Screenshot 2022-04-01 at 12 51 18" src="https://user-images.githubusercontent.com/275961/161258126-ee8690f2-0ab3-4ccb-b827-aec180bec2b9.png">

We should fix this in a different PR though...

mtias commented:
> If I have two navigation menus, one is displayed on the front end, but not in the site editor

How is this happening? It should be caught by the conditional.

scruffian commented:
> How is this happening? It should be caught by the conditional.

This PR doesn't touch the server side render, so I assume that's already the behaviour on the frontend.

youknowriad commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -143,6 +143,8 @@ function Navigation( {
 		);
 	}
 
+	const isCreatingEmpty = useRef( false );
```
If I'm understanding properly this ref is only needed because the user can create empty menus explicitly before the navigation menus load right? 

Can't we just hide all UI (keep the loader) while the navigation menus are being loaded?

getdave replied:
We need it because when the user clicks `Create empty` it clears out any existing ref`. At this point the block considers itself "empty" and thus tries to automatically utilise the first Navigation Menu if it exists. This means you can never get an empty menu cos it's always auto-filled with the first Nav Menu.

The flag exists simply to block that happening if the user has explicitly said "I want a blank Nav please". 

There are two states here:

1. Empty block that is inserted into canvas or already exists in canvas.
2. An empty menu that the user has created.

> Can't we just hide all UI (keep the loader) while the navigation menus are being loaded?

I tried that yes. It means the block takes longer for the Nav to reach a "settled" state. Moreover, that doesn't solve the issue I mentioned above because even if the block auto selects the first Nav Menu, sometimes as a user you want a entirely blank menu. Therefore you would click `Create empty` and because the block would already have the Nav Menu data available it would jump immediately to populating with the first one.

youknowriad replied:
Is it possible to infer that state (wanted empty menu) from state instead of using a ref?

The current implementation is a bit fragile and relying on the promise returned by data module actions in event callbacks is not a good practice. (React warning if the menu is removed before the action finishes)

getdave replied:
Thanks for following up here.

> The current implementation is a bit fragile and relying on the promise returned by data module actions in event callbacks is not a good practice.

It's certainly sub optimal. I can take a look again at whether we can infer from state but I seem to remember that wasn't possible.

The question I have is whether this is "good enough" (short term) to allow the functionality.

We can always follow up with a revised implementation post-6.0. But if you feel this is sufficiently brittle let's call it now and focus our efforts elsewhere. 

In the meantime I'll dig in again.


> React warning if the menu is removed before the action finishes

Yeh I've discussed this a few times with folks. The [last opinion given by @adamziel](https://github.com/WordPress/gutenberg/pull/38956#issuecomment-1047744390) was that this warning is being removed in React now anyway so let's not worry about it. I assume we're concerned about potential memory leaks?

getdave replied:
> The current implementation is a bit fragile and relying on the promise returned by data module actions in event callbacks is not a good practice.

@youknowriad I removed the reliance on `.finally()` and instead utilised the statuses I created in a previous PR to reset the flag.

I'd also like to understand more about why relying on the returned `Promise` is not considered to be a good practice.

adamziel replied:
@getdave I think that `isCreatingEmpty.current` could be replaced with `isSavingEntityRecord( 'postType', 'wp_navigation' )` without an id argument. The latter "Returns true if the specified entity record is saving, and false otherwise."

adamziel replied:
Hm although the ref is specific to the navigation block instance, whereas `isSavingEntityRecord` would flip true/false in all the blocks in the editor.

getdave replied:
Remember we have to detect whether the _user_ is creating an _empty_ menu. It's not just "we're creating a menu".

getdave replied:
Ok I've moved to using existing state rather than `useRef()`. Thanks for flagging this. It was a red flag there there might have been a simpler solution.

getdave committed:
Amend comment

Addresses https://github.com/WordPress/gutenberg/pull/39880#discussion_r840400402

getdave committed:
Avoid var declaration

Addresses https://github.com/WordPress/gutenberg/pull/39880#discussion_r840401238

getdave commented:
> > If I have two navigation menus, one is displayed on the front end, but not in the site editor
> 
> How is this happening? It should be caught by the conditional.

@scruffian Good catch 👏 

@mtias It's happening because of this line which tries to get an appropriate fallback if the block is empty:

https://github.com/WordPress/gutenberg/blob/2bdb3a45d5ab160654c163bf24ab30497cab56e5/packages/block-library/src/navigation/index.php#L455-L458

The fallback mechanic **for the front end** is (in order of precedence):

1. Try and find the first non-empty Navigation Menu and use that.
2. Use a list of your Pages.

https://github.com/WordPress/gutenberg/blob/2bdb3a45d5ab160654c163bf24ab30497cab56e5/packages/block-library/src/navigation/index.php#L328

This is what was agreed prior to WP 5.9. Happy to amend that code as necessary. What result should we look to achieve? Just ignore Nav Menus and use list of Pages?

getdave committed:
Avoid relying on promise returned from data actions

Addresses https://github.com/WordPress/gutenberg/pull/39880#discussion_r840697454

getdave committed:
Improve loading and placeholder conditionals around Nav Menus having been loaded

mtias commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -283,7 +297,22 @@ function Navigation( {
 		hasResolvedCanUserCreateNavigationMenu,
 	} = useNavigationMenu( ref );
 
+	// Attempt to retrieve and prioritize any existing navigation menu
```
Looks like we lost the longer comment. I think we should restore it:
```suggestion
	// Attempt to retrieve and prioritize any existing navigation menu unless
	// a specific ref is allocated or the user is explicitly creating a new menu. The aim is
	// for the block to "just work" from a user perspective using existing data.
```

getdave replied:
😄 I thought [you suggested to remove it](https://github.com/WordPress/gutenberg/pull/39880#discussion_r840400621)?

getdave replied:
Maybe I removed the wrong one? 👀

getdave replied:
Restored.

mtias commented:
@getdave that fallback for the front-end makes sense, we should just ensure we replicate on the back-end for empty menus not initiated by the user. We can follow up on that to not hold this one further.

getdave committed:
Restore comment

adamziel commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -283,7 +297,24 @@ function Navigation( {
+	// for the block to "just work" from a user perspective using existing data.
+	useEffect( () => {
+		if (
+			isCreatingEmpty.current ||
```
Does it have to be specifically `isCreatingEmpty`? As a user, I would expect this to short-circuit if this block's `isCreatingNavigationMenu` is true in general.

getdave replied:
🤔

getdave replied:
I'm sure I tried this approach. Let me see...

getdave replied:
This is very odd. I'm sure I tried that and there was a problem. I mean you're right it seems obvious.

I've just tried it again and now it seems to be working fine 🥳

getdave committed:
Fix loading state test

getdave committed:
Avoid ref and relying on existing state

Kudos to @adamziel https://github.com/WordPress/gutenberg/pull/39880#discussion_r841583225

getdave committed:
Remove duplicate e2e util

getdave committed:
Simplify by reverting additional handler function

draganescu commented on packages/e2e-tests/specs/editor/blocks/navigation.test.js:
```diff
@@ -1267,4 +1293,109 @@ describe( 'Navigation', () => {
+			);
+		} );
+
+		it( 'automatically uses the first Navigation Menu if only one available', async () => {
```
```suggestion
		it( 'automatically uses the first Navigation Menu if only one is available', async () => {
```

draganescu commented on packages/e2e-tests/specs/editor/blocks/navigation.test.js:
```diff
@@ -1267,4 +1293,109 @@ describe( 'Navigation', () => {
+			expect( linkText ).toBe( 'WordPress' );
+		} );
+
+		it( 'does not automatically use first Navigation Menu if be than one exists', async () => {
```
```suggestion
		it( 'does not automatically use first Navigation Menu if more than one exists', async () => {
```

draganescu reviewed (approved):
This looks good. I don't like that we need to wait for all the menus to load now before any interaction, maybe a better way is to preload menu data, the impact IMO should not be that big. However for the purposes of this PR the code looks good and it works as described.

getdave committed:
Fix test spellings

Co-authored-by: Andrei Draganescu <me@andreidraganescu.info>

mtias commented:
Thank you for working on this, I consider it an crucial step towards making menus more intuitive and reliable to use.

getdave commented:
> Thank you for working on this, I consider it an crucial step towards making menus more intuitive and reliable to use.

You are welcome 🙇
